File: src/main/java/rag/app/stages/IntentDetectionStage.java
package rag.app.stages;

import rag.app.Context;
import rag.app.StrategyRegistry;
import rag.intents.Intent;
import rag.intents.IntentDetector;
import rag.trace.TraceBus;
import rag.trace.TraceEvent;

public class IntentDetectionStage implements PipelineStage {

    private final IntentDetector detector;

    // StrategyRegistry DI (dependency injection)
    public IntentDetectionStage(StrategyRegistry registry) {
        this.detector = registry.getIntentDetector();
    }

    @Override
    public String getName() {
        return "IntentDetectionStage";
    }

    @Override
    public void run(Context context, TraceBus traceBus) throws Exception {

        long start = System.currentTimeMillis();
        String error = null;
        Intent detected = null;
        Exception failure = null;

        try {
            detected = detector.detect(context.getQuestion());
            context.setIntent(detected);
        } catch (Exception ex) {
            error = ex.getMessage();
            failure = ex;
        } finally {
            long duration = System.currentTimeMillis() - start;
            traceBus.publish(
                    new TraceEvent(
                            getName(),
                            "intent=" + detected,
                            duration,
                            error
                    )
            );
        }

        if (failure != null) {
            throw failure;
        }
    }
}

File: src/main/python/rag/app/stages/intent_detection_stage.py
from typing import Optional
from rag.app.stages.pipeline_stage import PipelineStage
from rag.app.context import Context
from rag.app.strategy_registry import StrategyRegistry
from rag.intents.intent import Intent
from rag.intents.intent_detector import IntentDetector
from rag.trace.trace_bus import TraceBus
from rag.trace.trace_event import TraceEvent
from rag.utils.timer_utils import now


class IntentDetectionStage(PipelineStage):
    def __init__(self, registry: StrategyRegistry) -> None:
        self.detector: IntentDetector = registry.get_intent_detector()

    def get_name(self) -> str:
        return "IntentDetectionStage"

    def run(self, context: Context, trace_bus: TraceBus) -> None:
        start: int = now()
        error: Optional[str] = None
        detected: Optional[Intent] = None
        failure: Optional[Exception] = None

        try:
            detected = self.detector.detect(context.get_question())
            context.set_intent(detected)
        except Exception as ex:
            error = str(ex)
            failure = ex
        finally:
            duration: int = now() - start
            trace_bus.publish(
                TraceEvent(
                    self.get_name(),
                    f"intent={detected}",
                    duration,
                    error
                )
            )

        if failure is not None:
            raise failure






